CRM/Financial/BAO/EntityFinancialAccount.php

```diff
@@ -45,28 +45,28 @@ public static function retrieve(&$params, &$defaults = [], &$allValues = []) {
    * @param array $params
    *   Reference array contains the values submitted by the form.
    * @param array $ids
-   *   Reference array contains one possible value
+   *   Deprecated array contains one possible value
    *   - entityFinancialAccount.
    *
    * @return CRM_Financial_DAO_EntityFinancialAccount
    * @deprecated
    * @throws \CRM_Core_Exception
    */
   public static function add(&$params, $ids = NULL) {
-    // action is taken depending upon the mode
-    $financialTypeAccount = new CRM_Financial_DAO_EntityFinancialAccount();
+    // action is taken depending upon the mode - which is wrong wrong wrong & should be like other add functions
     if ($params['entity_table'] !== 'civicrm_financial_type') {
+      $financialTypeAccount = new CRM_Financial_DAO_EntityFinancialAccount();
       $financialTypeAccount->entity_id = $params['entity_id'];
       $financialTypeAccount->entity_table = $params['entity_table'];
       $financialTypeAccount->find(TRUE);
+      $params['id'] = $financialTypeAccount->id;
     }
     if (!empty($ids['entityFinancialAccount'])) {
-      $financialTypeAccount->id = $ids['entityFinancialAccount'];
-      $financialTypeAccount->find(TRUE);
+      $params['id'] = $ids['entityFinancialAccount'];
+      CRM_Core_Error::deprecatedWarning('passing ids is deprecated');
     }
-    $financialTypeAccount->copyValues($params);
+    $financialTypeAccount = parent::writeRecord($params);
     self::validateRelationship($financialTypeAccount);
```

Comment:
@eileenmcnaughton it looks like these lines should be switched. The validation should happen before the record is written, right?

Reply:
oh dang - arg that looks hard to work around with the writeRecord

Reply:
I guess you could wrap it in a transaction and `rollback` if an exception is thrown?